### ISSUE TITLE
[BUG] Failing cudf v21.10 installation through conda

### ISSUE THREAD

**Issue by Niharikadutta** [rapidsai/cudf#9752]
**Describe the bug**
While trying to add `cudf=21.10` and `cudatoolkit=11.2` to an existing environment yaml file and creating an environment from that, it fails with the following system conflict error:

```
The following specifications were found to be incompatible with your system:

  - feature:/linux-64::__glibc==2.31=0
  - feature:|@/linux-64::__glibc==2.31=0
  - cudatoolkit=11.2 -> __glibc[version='>=2.17,<3.0.a0']
  - cudatoolkit=11.2 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - cudf=21.10 -> cupy[version='>7.1.0,<10.0.0a0'] -> __glibc[version='>=2.17|>=2.17,<3.0.a0']
  - gensim=3.8.3 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - hummingbird-ml=0.4.0 -> pytorch[version='>=1.4.*'] -> __glibc[version='>=2.17|>=2.17,<3.0.a0']
  - lightgbm=3.2.1 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - numpy=1.19.4 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - onnx=1.9.0 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - onnxruntime=1.7.2 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - pandas=1.2.3 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - pmdarima=1.8.2 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - pyarrow=3.0.0 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - pyodbc=4.0.30 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - python=3.8 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - pytorch=1.8.1 -> cudatoolkit[version='>=11.1,<11.2'] -> __glibc[version='>=2.17|>=2.17,<3.0.a0']
  - scikit-image=0.18.1 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - scikit-learn=0.23.2 -> libgcc-ng[version='>=7.5.0'] -> __glibc[version='>=2.17']
  - scipy=1.5.3 -> libgfortran-ng -> __glibc[version='>=2.17']
  - shap=0.39.0 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']
  - statsmodels=0.12.2 -> libgcc-ng[version='>=9.3.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.31

```
I'm not sure why it complains about glibc since version installed (2.31) satisfies the condition of `>=2.17`. Is there any guideline to installing cudf through conda on an existing environment? I've added all the recommended channels : `rapidsai`, `nvidia`, `conda-forge`.

**Expected behavior**
Installing cudf and cudatoolkit should succeed.

**Environment overview**
 - Environment location: Azure
 - Method of cuDF install: conda

**Comment by galipremsagar** [comment]
@Niharikadutta Is it possible to share the `yml` file to see the combination of packages you are trying to install?

**Comment by Niharikadutta** [comment]
Thanks @galipremsagar for looking into this.
Here is the `yml` file I'm I've added the following 2 packages to (removing them the environment builds successfully):

`cudf=21.10`
`cudatoolkit=11.2`

Resultant `yml` file:
```
name: Base
channels:
  - pytorch
  - interpretml
  - conda-forge
  - rapidsai
  - nvidia
  - defaults

dependencies:
  #
  # Core
  #
  - conda=4.9.2
  - python=3.8
  - ipython=7.23.1                 # Productive Interactive Computing
  - pip=21.1.1

  #
  # Core Python Data Science
  #
  - numpy=1.19.4            # Array processing for numbers, strings, records, and objects.
  - pandas=1.2.3            # High-performance, easy-to-use data structures and data analysis tools.
  - scikit-learn=0.23.2     # A set of python modules for machine learning and data mining
  - sklearn-pandas=2.2.0    # Pandas integration with sklearn
  - scikit-image=0.18.1     # Image processing in Python.
  - scipy=1.5.3             # Scientific Library for Python
  - pyarrow=3.0.0           # Python libraries for Apache Arrow
  - matplotlib=3.4.2        # Publication quality figures in Python

  #
  # Data Science libraries
  #
  - seaborn=0.11.1          # Statistical data visualization
  - bokeh=2.3.2             # Statistical and novel interactive HTML plots for Python
  - nltk=3.6.2              # Natural Language Toolkit
  - pmdarima=1.8.2          # Time series analysis capabilities
  - patsy=0.5.1             # Describing statistical models in Python using symbolic formulas
  - plotly=4.14.3           # An interactive, browser-based graphing library for Python
  - statsmodels=0.12.2      # Statistical computations and models for use with SciPy
  - textblob=0.15.3         # Simple, Pythonic text processing.
  - mleap=0.17.0            # MLeap Python API
  - mlflow-skinny=1.18.0    # MLflow is an open source platform for the machine learning lifecycle.
  - gensim=3.8.3            # Topic Modelling for Humans

  #
  # Utilities
  #
  - geopy=2.1.0             # Python Geocoding Toolbox.
  - html5lib=1.1            # HTML parser based on the WHATWG HTML specification
  - beautifulsoup4=4.9.3    # Python library designed for screen-scraping
  - imageio=2.9.0           # A Python library for reading and writing image data
  - isodate=0.6.0           # An ISO 8601 date/time/duration parser and formatter.
  - markdown=3.3.4          # Python implementation of Markdown.
  - openpyxl=3.0.7          # A Python library to read/write Excel 2010 xlsx/xlsm files
  - urllib3=1.26.4          # HTTP library with thread-safe connection pooling, file post, and more.
  - liac-arff=2.5.0         # A module for read and write ARFF files in Python.


  #
  # Azure Data Sources
  #
  - adlfs=0.7.7             
  - pyodbc=4.0.30       

  #
  # PySpark
  #
  - pyspark=3.1.2          # Apache Spark
  - koalas=1.8.0           # pandas API on Apache Spark

  #
  # PyTorch and friends
  #
  - pytorch=1.8.1           # PyTorch is an optimized tensor library for deep learning using GPUs and CPUs. DEPENDENCY: PyTorch Enterprise OWNER: PyTorch Enterprise

  #
  # Trees
  #
  - lightgbm=3.2.1                # Gradient boosting framework 

  #
  # Hummingbird
  #
  - hummingbird-ml=0.4.0          # Convert trained traditional machine learning models into tensor computations.

  #
  # ONNX
  #
  - onnx=1.9.0                    # Open Neural Network Exchange library
  - onnxmltools=1.7.0             # enables conversion of models to ONNX
  - keras2onnx=1.6.5              # Convert tf.keras/Keras models to ONNX
  - skl2onnx=1.8.0.1              # Convert scikit-learn models and pipelines to ONNX
  - onnxruntime=1.7.2             # A runtime accelerator for Machine Learning models; DEPENDENCY: TOUCAN, SQL DW

  #
  # Explainable ML
  #
  - shap=0.39.0                    # A unified approach to explain the output of any machine learning model.
  - interpret=0.2.4                # Fit interpretable models. Explain blackbox machine learning.

  #
  # RAPIDS libraries
  #
  - cudf=21.10
  - cudatoolkit=11.2

  - pip:

    - azureml-mlflow==1.34.0       
    - azureml-core==1.34.0
    - azureml-opendatasets==1.34.0
    #
    # PyTorch and friends
    #
    - torchvision==0.9.1           # PyTorch is an optimized tensor library for deep learning using GPUs and CPUs. DEPENDENCY: PyTorch Enterprise OWNER: PyTorch Enterprise 

    #
    # TensorFlow
    #
    - tensorflow==2.4.1              # TensorFlow is an open source machine learning framework for everyone.
    - tensorboard==2.4.1             # TensorFlow's Visualization Toolkit
    - tensorflow-estimator==2.4.0    # TensorFlow Estimator is a high-level TensorFlow API that greatly simplifies machine learning programming.
    - keras-applications==1.0.8      # Applications module of the Keras deep learning library.
    - keras-preprocessing==1.1.2     # Data preprocessing and data augmentation module of the Keras deep learning library

    #
    # Not on conda-forge
    #
    - pandasql==0.7.3              # sqldf for pandas

    #
    # Other python dependencies in BBC-VHD
    #
    - mypy==0.780
    - mypy-extensions==0.4.3
    - typed-ast==1.4.3
```

**Comment by galipremsagar** [comment]
Thanks for the `yml` file @Niharikadutta. The issue is because you have conflicting version requirements for `pyarrow`, the yml file requires `pyarrow=3.0.0` but `cudf=21.10` requires `pyarrow=5.0.0` to be present. So a solution to fix this is to comment out the `pyarrow=3.0.0` and it should create you an environment. But there is an additional issue with the pip packages you are installing in this `yml` file. The `azureml-core==1.34.0` seems to remove the `pyarrow=5.0.0` installed from `conda-forge` and replace it with `pyarrow=3.0.0` from `pypi`. This will result in a binary mismatch for any conda package that relies on `pyarrow` installed from `conda-forge`. I'm not sure what the use-case and alternatives here are but removing those pypi `azureml` packages from `yml` will avoid the binary size mismatch issue.

**Comment by galipremsagar** [comment]
Closing this issue as it is resolved from a `cudf` standpoint.
